pydra/engine/specs.py

```diff
@@ -48,18 +48,33 @@ def hash(self):
                 "output_file_template"
             ):
                 continue
-            if field.type == File and "container_path" not in field.metadata:
-                inp_dict[field.name] = hash_file(getattr(self, field.name))
-            elif isinstance(getattr(self, field.name), tuple):
-                inp_dict[field.name] = list(getattr(self, field.name))
-            else:
-                inp_dict[field.name] = getattr(self, field.name)
+            inp_dict[field.name] = self.hash_value(
+                value=getattr(self, field.name),
+                type=field.type,
+                metadata=field.metadata,
+            )
         inp_hash = hash_function(inp_dict)
         if hasattr(self, "_graph_checksums"):
             return hash_function((inp_hash, self._graph_checksums))
         else:
             return inp_hash
 
+    def hash_value(self, value, type=None, metadata=None):
+        """calculating hash or returning values recursively"""
+        from .helpers_file import hash_file
+
+        if isinstance(value, (tuple, list)):
+            return [self.hash_value(el, type, metadata) for el in value]
```

Comment:
shouldn't something be inspecting the type of `el`?

Reply:
it should be check when the `hash_value` is called

Reply:
if `el` is not container anymore, it should go to line 71